File: VSYard/VSYard/Classifiers/YardClassifier.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.ComponentModel.Composition;
using Microsoft.VisualStudio.Text;
using Microsoft.VisualStudio.Text.Classification;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.Text.Tagging;
using Microsoft.VisualStudio.Utilities;


namespace MyCompany.VSYard.Classifiers
{
    [Export(typeof(ITaggerProvider))]
    [ContentType("yard")]
    [TagType(typeof(ClassificationTag))]
    internal sealed class YardClassifierProvider : ITaggerProvider
    {
        [Export]
        [Name("yard")]
        [BaseDefinition("code")]
        internal static ContentTypeDefinition OokContentType = null;

        [Export]
        [FileExtension(".yrd")]
        [ContentType("yard")]
        internal static FileExtensionToContentTypeDefinition OokFileType = null;

        [Import]
        internal IClassificationTypeRegistryService ClassificationTypeRegistry = null;

        [Import]
        internal IBufferTagAggregatorFactoryService aggregatorFactory = null;

        public ITagger<T> CreateTagger<T>(ITextBuffer buffer) where T : ITag
        {

            ITagAggregator<VSYardNS.TokenTag> ookTagAggregator =
                                            aggregatorFactory.CreateTagAggregator<VSYardNS.TokenTag>(buffer);

            return new VSYardNS.YardClassifier(buffer, ookTagAggregator, ClassificationTypeRegistry) as ITagger<T>;
        }

    }
    /*internal sealed class YardClassifier : ITagger<ClassificationTag>
    {
        ITextBuffer _buffer;
        ITagAggregator<VSYardNS.TokenTag> _aggregator;
        IDictionary<VSYardNS.TokenTypes, IClassificationType> _ookTypes;

        internal YardClassifier(ITextBuffer buffer,
                               ITagAggregator<VSYardNS.TokenTag> ookTagAggregator,
                               IClassificationTypeRegistryService typeService)
        {
            _buffer = buffer;
            _aggregator = ookTagAggregator;
            _ookTypes = new Dictionary<VSYardNS.TokenTypes, IClassificationType>();
            _ookTypes[VSYardNS.TokenTypes.TLitersl] = typeService.GetClassificationType("ook!");
            _ookTypes[VSYardNS.TokenTypes.TNonterm] = typeService.GetClassificationType("ook.");
            _ookTypes[VSYardNS.TokenTypes.TTerm] = typeService.GetClassificationType("ook?");
        }

        public event EventHandler<SnapshotSpanEventArgs> TagsChanged
        {
            add { }
            remove { }
        }

        public IEnumerable<ITagSpan<ClassificationTag>> GetTags(NormalizedSnapshotSpanCollection spans)
        {

            foreach (var tagSpan in this._aggregator.GetTags(spans))
            {
                var tagSpans = tagSpan.Span.GetSpans(spans[0].Snapshot);
                yield return
                    new TagSpan<ClassificationTag>(tagSpans[0],
                                                   new ClassificationTag(_ookTypes[tagSpan.Tag.Type]));
            }
        }
    }*/

}

File: VSYard/VSYardLibrary/VSYardLibrary/YardClassifier.fs
namespace VSYardModule
open System
open System.Collections.Generic
open System.ComponentModel.Composition
open Microsoft.VisualStudio.Text
open Microsoft.VisualStudio.Text.Classification
open Microsoft.VisualStudio.Text.Editor
open Microsoft.VisualStudio.Text.Tagging
open Microsoft.VisualStudio.Utilities
open System.Linq;

[<Export(typeof<ITaggerProvider>)>]
[<ContentType("yard")>]
[<TagType(typeof<ClassificationTag>)>]
type internal YardClassifier () = class
        [<Export>]
        [<Name("yard")>]
        [<BaseDefinition("code")>]
        let mutable ookContentType : ContentTypeDefinition = null
        member self.OokContentType 
            with get () = ookContentType
            and set oot = ookContentType <- oot

        [<Export>]
        [<FileExtension(".yrd")>]
        [<ContentType("yard")>]
        let mutable ookFileType : FileExtensionToContentTypeDefinition = null
        member self.OokFileType 
            with get () = ookContentType
            and set oft = ookContentType <- oft

        [<Import>]
        let mutable classificationTypeRegistry : IClassificationTypeRegistryService = null
        member self.ClassificationTypeRegistry
            with get () = classificationTypeRegistry
            and set ctr = classificationTypeRegistry <- ctr

        [<Import>]
        let mutable aggregatorFactory : IBufferTagAggregatorFactoryService = null
        member self.AggregatorFactory 
            with get () = aggregatorFactory
            and set af = aggregatorFactory <- af

        interface ITaggerProvider with
            member this. CreateTagger<'T when 'T :> ITag>(buffer : ITextBuffer) : ITagger<'T> =
                let ookTagAggregator : ITagAggregator<TokenTag> = aggregatorFactory.CreateTagAggregator<TokenTag>(buffer)
                MyClassifier(buffer, ookTagAggregator, ClassificationTypeRegistry) :> ITagger<'T>; 
    end
type internal MyClassifier () = class
    interface ITagger<ClassificationTag>
    end